### PR TITLE
feat: implement incremental merkle tree

### PR DESCRIPTION
**Pull request by YashBit**
## Important:

1. Creating a simple PR for just a feedback loop, code not ready yet. Will update the description, etc when I have made all the commits. 
2. Want to check if I am on the correct path and implementing all the necessary components.

### CODE REVIEW

**Comment by YashBit** [comment]
@cedoor I have created a DynamicCalculator trait which will works the same way but with the depth as a parameter now. This basically controls the loop. 

In -> packages/merkle-trees/src/incremental_merkle/tree.nr

Please let me know if this is the correct direction of implementation with a simple new modified trait that I have completed. 

After which I will focus on the MembershipProver, etc methods. Then write the Poseidon Hash tests.

**Review by cedoor — COMMENTED** [review]
@YashBit there's code which is not related to this PR (e.g. ecdh). I think you need to open a PR for the binary merkle tree by forking the main branch of zk-kit.noir.

File: packages/merkle-trees/src/incremental_merkle.nr
Comment (cedoor) [inline packages/merkle-trees/src/incremental_merkle.nr:1]: This file looks the same as: https://github.com/privacy-scaling-explorations/zk-kit.noir/blob/main/packages/merkle-trees/src/merkle.nr . I think you can re-use the same code. In the end what you should implement here is the same tree (binary merkle tree, the 'incremental' property is out of context here).

File: packages/merkle-trees/src/lib.nr [hunk 1]
```diff
@@ -57,6 +66,27 @@ trait MT_Creator {
...
+
+    /**
+     * Creates a new Incremental Merkle Tree (IncrementalMerkleTree) instance.
+     * @param hasher The hash function that is used to hash the nodes of the tree
+     */
+    fn new(hasher: fn([Field; 2]) -> Field) -> Self {
+        Self::from(0, hasher)
+    }
+}
+
```
Comment (cedoor) [inline packages/merkle-trees/src/lib.nr:89]: This is also the same as the existing `trait MT_Creator`. I guess you can re-use the same code.

**Review by cedoor — CHANGES_REQUESTED** [review]

**Comment by cedoor** [comment]
@YashBit 

>  1. Please let me know if this is the correct direction of implementation with a simple new modified trait that I have completed.

The trait seems to be fine, I added other comments for some files that are actually just a copy of the current Merkle tree. You should be able to re-use that code.
 
>  2. After which I will focus on the MembershipProver, etc methods. Then write the Poseidon Hash tests.
 
 Yes, those functions still need to be implemented I guess.
 
> For the tests, how would you like me to test the dynamic depth function? So I should create a tree (let's say of depth 4), and then calculate the results for depth 1, 2, 3?

Exactly, sounds like a good way to test them. Please, take a look at these tests here: https://github.com/privacy-scaling-explorations/zk-kit.circom/blob/326cef9fdb9a2f845b890cffea0594975768be1f/packages/binary-merkle-root/tests/binary-merkle-root.test.ts#L98. You should have the same tests in Noir here.

Also, be aware that this is not an incremental Merkle tree. It works as a general binary Merkle tree with an addition feature: dynamic depth, i.e. the ability to generate a proof with an arbitrary depth even if there's a maximum depth in the circuit (which must be static by design).

**Comment by signorecello** [comment]
Converting to draft while not ready to merge

File: package.json [hunk 2]
```diff
@@ -27,10 +27,21 @@
   "devDependencies": {
     "@commitlint/cli": "^18.6.1",
     "@commitlint/config-conventional": "^18.6.2",
+    "@types/chai": "^4",
+    "@types/mocha": "^10",
```
Comment (sripwoud) [inline package.json:31]: are you planning to write some node tests for this too (with mocha) like in #8?
If not, we don't need these deps

**Comment by YashBit** [comment]
@signorecello Please can you change this back from the draft? The core logic is complete, adding a few tests now.

**Comment by YashBit** [comment]
@signorecello 

Could you please comment on the algorithm? Most of it should remain the same, it is just the calculate_root function in now a new trait called DynamicCalculator which should be able to handle the depth function dynamically.

Thank you again.

**Review by sripwoud — CHANGES_REQUESTED** [review]
@YashBit I think you need to [`rebase`](https://git-scm.com/docs/git-rebase) your branch on `main`. This PR shows up changes related to eg mocha dependencies while you added these in #8  already.  

And formatting CI check is still failing

**Comment by YashBit** [comment]
@sripwoud I will make those changes.

I am mainly concerned with the logic of the incremental tree. That is the most important, since after that I will be done with unit the tests as well.

**Comment by signorecello** [comment]
Hey @YashBit I don't understand the goal here. When you use this library in a circuit, you won't be able to pass anything other than an array into its `main` circuit.

I don't believe we should merge this PR without a `bin` circuit that would show its usage

**Review by signorecello — CHANGES_REQUESTED** [review]
Hey @YashBit I don't understand the goal here. When you use this library in a circuit, you won't be able to pass anything other than an array into its main circuit.

I don't believe we should merge this PR without a bin circuit that would show its usage

**Comment by cedoor** [comment]
@YashBit can we close this PR in favor of https://github.com/privacy-scaling-explorations/zk-kit.noir/pull/16 ?

**Comment by YashBit** [comment]
@cedoor Yes, please close this one!

**Pull request closed without merging**